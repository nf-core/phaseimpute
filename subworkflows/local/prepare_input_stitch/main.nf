include { GAWK                      } from '../../../modules/nf-core/gawk'

workflow PREPARE_INPUT_STITCH {

    take:
    ch_input_impute  // channel:   [ [id], bam, bai ]
    ch_posfile       // channel:   [ [panel, chr], legend ]
    ch_region        // channel:   [ [chr, region], region ]

    main:

    ch_versions      = Channel.empty()

    // Value channels
    def input_empty         = [[]]
    def rdata_empty         = [[]]
    k_val                   = params.k_val
    ngen                    = params.ngen

    // Transform posfile to TSV with ','
    GAWK(ch_posfile, [])

    // Get chromosomes of posfile
    ch_posfile = GAWK.out.output
        .map{metaPC, posfile -> [[chr: metaPC.chr], metaPC, posfile]}

    // Get chromosomes of fasta
    ch_chromosomes = ch_region
        .map{metaCR, region -> [[chr: metaCR.chr], metaCR.chr]}

    // Make final channel with parameters
    stitch_parameters = ch_posfile
        .map { it + input_empty + rdata_empty}
        .join(ch_chromosomes)
        .map { it + k_val + ngen}
        .map { metaC, metaPC, posfile, input, rdata, chr, k_val, ngen ->
            [metaPC, posfile, input, rdata, chr, k_val, ngen]
        }

    // Prepare sample files for STITCH
    // Group input by ID
    ch_bam_bai = ch_input_impute
        .map {metaI, bam, bai -> [metaI.subMap("id"), bam, bai]}
        .unique()

    // Make bamlist from bam input
    ch_bamlist = ch_bam_bai
        .map {it[1].toString().tokenize('/').last()}
        .collectFile(name: "bamlist.txt", newLine: true, sort: true)

    // Collect all files
    stitch_samples = ch_bam_bai
        .map {meta, bam, bai -> [[id: "all_samples"], bam, bai]}
        .groupTuple()
        .combine(ch_bamlist)
        .collect()

    emit:
    stitch_parameters = stitch_parameters // channel:   [ [chr], posfile, [], [], chr, k_val, ngen ]
    stitch_samples    = stitch_samples    // channel:   [ [id], bam, bai, bamlist ]
    versions          = ch_versions       // channel:   [ versions.yml ]
}
