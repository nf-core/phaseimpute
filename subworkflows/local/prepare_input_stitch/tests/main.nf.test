nextflow_workflow {

    name "Test Subworkflow PREPARE_INPUT_STITCH"
    script "../main.nf"
    config "./nextflow.config"

    workflow "PREPARE_INPUT_STITCH"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/prepare_input_stitch"
    tag "prepare_input_stitch"

    tag "stitch"

    test("Prepare input stitch") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "NA12878"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA12878/NA12878.s.bam.bai", checkIfExist:true),
                    ],
                    [
                        [id: "NA19401"],
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/individuals/NA19401/NA19401.s.bam.bai", checkIfExist:true),
                    ],
                ])
                input[1] = Channel.of([
                        [id: "1000GP", chr: "22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.legend.gz", checkIfExist:true)
                    ]).collect()
                input[2] = Channel.of([
                    [chr: "22", region: "chr22:16570000-16610000"], "chr22:16570000-16610000"
                ]).collect()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
