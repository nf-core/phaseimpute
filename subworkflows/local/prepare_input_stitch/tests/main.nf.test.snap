{
    "Prepare input stitch": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "22"
                        },
                        "1000GP_22_posfile_stitch.txt:md5,485f95911c84e9de54c6ec9eda26b385",
                        [
                            
                        ],
                        [
                            
                        ],
                        "22",
                        2,
                        100
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "all_samples"
                        },
                        [
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA12878/NA12878.s.bam",
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA19401/NA19401.s.bam"
                        ],
                        [
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA12878/NA12878.s.bam.bai",
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA19401/NA19401.s.bam.bai"
                        ],
                        "bamlist.txt:md5,70f6e727cdcb642c1329cca4f5d77775"
                    ]
                ],
                "2": [
                    
                ],
                "stitch_parameters": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "22"
                        },
                        "1000GP_22_posfile_stitch.txt:md5,485f95911c84e9de54c6ec9eda26b385",
                        [
                            
                        ],
                        [
                            
                        ],
                        "22",
                        2,
                        100
                    ]
                ],
                "stitch_samples": [
                    [
                        {
                            "id": "all_samples"
                        },
                        [
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA12878/NA12878.s.bam",
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA19401/NA19401.s.bam"
                        ],
                        [
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA12878/NA12878.s.bam.bai",
                            "/nf-core/test-datasets/phaseimpute/hum_data/individuals/NA19401/NA19401.s.bam.bai"
                        ],
                        "bamlist.txt:md5,70f6e727cdcb642c1329cca4f5d77775"
                    ]
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-17T16:25:05.261820347"
    }
}